source/extensions/transport_sockets/tls/ssl_socket.cc

```diff
@@ -379,13 +379,41 @@ bool ClientSslSocketFactory::implementsSecureTransport() const { return true; }
 
 void ClientSslSocketFactory::onAddOrUpdateSecret() {
   ENVOY_LOG(debug, "Secret is updated.");
+  bool should_run_callbacks = false;
   {
     absl::WriterMutexLock l(&ssl_ctx_mu_);
     ssl_ctx_ = manager_.createSslClientContext(stats_scope_, *config_, ssl_ctx_);
+    if (ssl_ctx_) {
+      should_run_callbacks = true;
+    }
+  }
+
+  if (should_run_callbacks) {
+    absl::WriterMutexLock m(&secrets_ready_callbacks_mu_);
+    for (const auto& cb : secrets_ready_callbacks_) {
+      cb();
+    }
+    secrets_ready_callbacks_.clear();
   }
   stats_.ssl_context_update_by_sds_.inc();
 }
 
+void ClientSslSocketFactory::addReadyCb(std::function<void()> callback) {
+  bool immediately_run_callback = false;
+  {
+    absl::ReaderMutexLock l(&ssl_ctx_mu_);
+    if (ssl_ctx_) {
+      immediately_run_callback = true;
+    } else {
+      absl::WriterMutexLock m(&secrets_ready_callbacks_mu_);
+      secrets_ready_callbacks_.push_back(callback);
+    }
+  }
+  if (immediately_run_callback) {
+    callback();
+  }
+}
```

Comment:
Consider this change here and in `ServerSslSocketFactory::addReadyCb()`. This way the ssl_ctx lock could be released earlier.

```suggestion
void ClientSslSocketFactory::addReadyCb(std::function<void()> callback) {
  bool immediately_run_callback = false;
  {
    absl::ReaderMutexLock l(&ssl_ctx_mu_);
    if (ssl_ctx_) {
      immediately_run_callback = true;
    }
  }
  if (immediately_run_callback) {
    callback();
  } else {
    absl::WriterMutexLock m(&secrets_ready_callbacks_mu_);
    secrets_ready_callbacks_.push_back(callback);
  }
}
```

Reply:
don't I need to hold both locks when adding to the callbacks list? otherwise another thread could set `ssl_ctx_mu` and then not see this callback should be run